### ISSUE TITLE
Allow for creation of groups corresponding to Sections

### ISSUE THREAD

**Issue by klemay** [hypothesis/lms#419]
As an instructor of a course with 20+ students (or maybe even 100+ students!), I want students to read and annotate in small groups. If all groups are annotating the same document, I wouldn't want students seeing annotations made by other groups. 

## User requests
- https://hypothesis.zendesk.com/agent/tickets/3582
- https://hypothesis.zendesk.com/agent/tickets/3637
- this also came up during a call with a key potential edu partner last week
- https://hypothesis.zendesk.com/agent/tickets/3724
- https://hypothesis.zendesk.com/agent/tickets/3733

**Comment by jeremydean** [comment]
I believe the LTI spec supports this by allowing for "context" to be a "group" (or section) rather than a "course" as a whole.

https://www.imsglobal.org/specs/ltiv1p0/implementation-guide

**Comment by seanh** [comment]
@jeremydean @klemay So it's possible to create a "group within a course" in Canvas? How exactly is that done in the Canvas UI? That would be helpful to know in order to assess the possibilities here.

Note that we've already implemented per-course groups though. Per-group-within-a-course groups is a different and conflicting feature and while some users might want it, if we implemented per-group-within-a-course groups would we then get complaints from other users who preferred the simpler per-course groups? Having multiple groups within a course seems like it would also complaint the activity overviews and grading views that we've been discussing.

So this seems like something that would require plenty of thought and discussion, rather than something to do quickly

**Comment by jeremydean** [comment]
For clarity, I was imagining instructor users would be given a choice initially to _either_ have a single Hypothesis group for the course or multiple Hypothesis groups for sections within the course. 

I have heard that it'd be nice to have both at the same time, moving from one to the other for different assignments, but as a first step it's definitely either/or with most big classes likely choosing multiple groups for sections and smaller classes just fine with the course=group. 

So, sub-groups might not be the best term here. I'll follow up with screenshots for group creation in Canvas, etc.

**Comment by ajpeddakotla** [comment]
Spoke with @jeremydean this afternoon about this, and I think there's been a mixup here (or at least I was mixing things up). The problem that we're trying to solve with this issue is the following: 
Teachers need to be able to divide the class into smaller cohorts, so that a 100 person class for example can be divided into 5 groups of 20 people each. 

The decision to create small cohorts of a class would be up to an instructor. The terminology for this "cohort" needs to be decided on, as well as the user flows for the creation of this "cohort". Wether this group is a "sub-group" or just another group associated with the course is TBD.

**Comment by klemay** [comment]
For context, within the LMS, when you break students up into smaller sections, that is generally referred to as a "group" or "section."

"Section" is used when an instructor teaches, for example, Chem 101 on Monday-Wednesday and Tuesday-Thursday. The M-W students would be in "Section 1" and the T-R students would be in "Section 2." LMSes allow that instructor to have both Sections in the same course site so they don't have to upload materials in 2 places. These Sections are tied to the roster, so students are automatically sorted. 

"Groups" are chosen by the instructor - think of them as manually-created "Sections" that are decided by the instructor and not tied to the roster. So if I'm teaching English 101 with 20 students, I could manually create 5 groups of 4 students.

Just to make things extra fun, depending on the LMS, Sections and Groups are sometimes merged into 1 feature. For example, in Sakai, if I'm teaching that Chem 101 class, I could:

- Automatically create groups for Section 1 and Section 2 based on the roster
- Create additional groups within my sections, breaking students down into smaller groups

...so when I'm creating an assignment, I can say "assign this to Section 1" or "assign this to Group 5, Section 1". 

I said "sub-group" in my initial issue title only because right now a group=course, and I wanted to be clear that users were asking for smaller divisions within the course site. I'm not at all attached to the nomenclature.

**Comment by dwhly** [comment]
Super helpful, thank you @klemay!

**Comment by seanh** [comment]
It's gonna be really helpful for us if the LMS's send us an ID for the section/group in the launch params, because then we could use that to automatically create different groups for different sections.

I don't _think_ there's a standard LTI launch param for this. The LTI launch parameters contain the concept of a "context" (`context_id`, `context_type`, `context_title`, `context_label`) but I think in practice the "context" here maps to the "course" rather than to a group or section within the course, because we _already_ create a different group for each context. So unless we already do per-section groups, then context = course!

Maybe there's a non-standard launch param that LMS's send us that we can use here. Would need to take a close look at the launch params from actual LMS's.

If there's nothing in the launch params that we can use then things might be more difficult.

**Comment by klemay** [comment]
From the [implementation docs](https://www.imsglobal.org/specs/ltiv1p0/implementation-guide#toc-14) Jeremy linked to above, some of the available `context_type` parameters are:

CourseSection
Group

**Comment by seanh** [comment]
Yep, but I suspect that LMS's don't actually support that part of the spec. According to <https://www.imsglobal.org/toolcompare.html> D2L does and Sakai might but not Canvas, Blackboard or Moodle. We'll investigate later and find out for sure.

(That doc is the LTI 1.0 spec by the way, "Implementation Guide" is what IMS Global calls specifications.)

**Comment by klemay** [comment]
Got it, thanks for the clarification!

**Comment by seanh** [comment]
I'm saying that if LMS's don't actually send us the launch params to tell us which course section or group is being used, then we'll have to do something much more difficult to try to find out that information. But, hopefully on investigation this info will turn out to be in the launch params somewhere.

**Comment by mkdir-washington-edu** [comment]
Another ask for Canvas Groups functionality: https://hypothesis.zendesk.com/agent/tickets/6022
tl;dr User has 50 students, got up to 380 annotations in one article. Would like to separate assignments among students in 5 to 6 groups.

**Comment by mkdir-washington-edu** [comment]
User request that I think best fits with this issue: https://hypothesis.zendesk.com/agent/tickets/6450

The user wants to be able to assign a single text to multiple groups within a course, and then give the instructor the option of seeing either the section's work or seeing the course's aggregate work. This will allow the instructor to easily see similarities between the locations and content of the different groups' annotations.

**Comment by jeremydean** [comment]
I don't know if this is the right place to add a wrinkle, but I've discovered two different contexts in which courses can be sectioned:

1) is as described above: I have a class of 100 students and have divided them into "groups" within the LMS.

2) is different: I teach 4 sections of the same course that use the same course shell. I believe this sectioning is done differently then in 1) above. From what I've heard from users, Hypothesis is not recognizing these different classes as distinct groups.  

CC @klemay

**Comment by klemay** [comment]
Thanks for flagging this, @jeremydean! We discussed this a bit above, but it's worth revisiting as we get closer to being able to work on this. 

We need to determine:
- Are we building for scenario 1 and 2 in your comment? Just 1? Just 2? 
- If we're building for both, which comes first? 

Scenario 1 (I have a class of X number of students and have divided them into "groups" within the LMS) is the one we get feature requests for the most often, but what's your take here?

**Comment by mkdir-washington-edu** [comment]
Is 2 solvable by implementing 1? I.e., could each section be put into a group, thereby keeping each sections annotations distinct?

**Comment by mkdir-washington-edu** [comment]
Also: https://hypothesis.zendesk.com/agent/tickets/7118

**Comment by klemay** [comment]
To sum up a conversation with Jeremy & engineers that happened earlier this week:

## There are multiple ways of creating "groups" in an LMS:

Automatically based on Sections, derived from the school's Student Information system. This covers a few scenarios:

- I teach English 101 with 100 students every Tuesday and Thursday. They are divided into 4 Sections of 25 each, who meet with a TA outside of lecture. So I have English 101 Section 1, English 101 Section 2... etc). They are all in the same course site because they read the same materials, complete the same assignments, etc. 
- I teach Physics 101 to 2 Sections of 25 students. Section 1 meets on Monday and Wednesday, and Section 2 meets on Tuesday and Thursday. While they don't sit in the classroom on the same days, Sections 1 and 2 read the same materials, complete the same assignments, etc., though their assignments have different due dates. I can make homework for Section 1 due on Wednesday and homework for Section 2 due on Thursday, for example. 
- I teach Chemistry 101 along with 2 other colleagues. I teach 25 students on Monday and Wednesday at 9am, one colleague teaches 25 students on Monday and Wednesday at 12pm, and the other teaches 25 students on Tuesday and Thursday at 9am. The department sets our curriculum, so the three of us share one course site with all of our students, where materials are pre-populated. Assignments likely have different due dates for each section. 

Manually through a tool built into the LMS:

- I teach English 404 to 20 students. I like for them to work in small groups, so I use Canvas' Groups tool to create 5 groups of 4 students. I can assign a different poem to each group. 

## LMSes have differing groups capabilities

Some LMSes allow for all of the group creation methods above. Others allow for only the Sections method but no manual group creation. 

## Some LTI tools forgo the LMS groups for their own tool

Rather than trying to determine how each LMS handles groups and working with that, some tools will build in their own groups creation capability. So I could assign a Hypothesis reading to a select few students, not through SIS data or the LMS' groups creation possibility, but through a feature Hypothesis would design and build

## LTI 1.3 will likely require groups capability

The [LTI 1.3 spec](https://www.imsglobal.org/spec/lti/v1p3#overview-0) does mention Sections and Groups, and we will be pursuing 1.3 certification down the road. If certification requires us to support groups and/or sections in a way that would meet our users' goals, we should do it according to the spec!

## Research will be required to determine a path forward

**Comment by seanh** [comment]
Doc about sections: https://docs.google.com/document/d/1sb7ymaOsoDL16M980y-ADfZr8NY9VBv9l40E2dvSWuc/

**Comment by seanh** [comment]
Closing in favour of <https://github.com/hypothesis/lms/issues/1296>. Note that <https://github.com/hypothesis/lms/issues/1296> is Canvas-specific. Please open separate issues for the equivalent features in other LMS's
